mempool/src/pool.rs

```diff
@@ -174,44 +189,49 @@ impl RollingFeeRate {
     }
 }
 
-#[derive(Debug)]
 pub struct Mempool<
-    C: ChainState + 'static + Send,
-    H: 'static + Send,
     T: GetTime + 'static + Send,
-    M: GetMemoryUsage + 'static + Send,
+    M: GetMemoryUsage + 'static + Send + std::marker::Sync,
 > {
+    #[allow(unused)]
+    chain_config: Arc<ChainConfig>,
     store: MempoolStore,
-    rolling_fee_rate: Cell<RollingFeeRate>,
+    rolling_fee_rate: RwLock<RollingFeeRate>,
     max_size: usize,
     max_tx_age: Duration,
-    chain_state: C,
-    #[allow(unused)]
-    chainstate_handle: H,
+    chainstate_handle: subsystem::Handle<Box<dyn ChainstateInterface>>,
     clock: T,
     memory_usage_estimator: M,
 }
 
-impl<C, H, T, M> Mempool<C, H, T, M>
+impl<T, M> std::fmt::Debug for Mempool<T, M>
 where
-    C: ChainState + Send,
-    H: Send,
-    T: GetTime + Send,
-    M: GetMemoryUsage + Send,
+    T: GetTime + 'static + Send + std::marker::Sync,
+    M: GetMemoryUsage + 'static + Send + std::marker::Sync,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self.store)
+    }
+}
+
+impl<T, M> Mempool<T, M>
+where
+    T: GetTime + Send + std::marker::Sync,
+    M: GetMemoryUsage + Send + std::marker::Sync,
 {
     pub(crate) fn new(
-        chain_state: C,
-        chainstate_handle: H,
+        chain_config: Arc<ChainConfig>,
+        chainstate_handle: subsystem::Handle<Box<dyn ChainstateInterface>>,
         clock: T,
         memory_usage_estimator: M,
     ) -> Self {
         Self {
+            chain_config,
             store: MempoolStore::new(),
-            chain_state,
             chainstate_handle,
             max_size: MAX_MEMPOOL_SIZE_BYTES,
             max_tx_age: DEFAULT_MEMPOOL_EXPIRY,
-            rolling_fee_rate: Cell::new(RollingFeeRate::new(clock.get_time())),
+            rolling_fee_rate: parking_lot::RwLock::new(RollingFeeRate::new(clock.get_time())),
```

Comment:
I'm not sure we really have to use parking_lot anymore. In a recent version of rust, mutexes have been improved to solve the performance problems that parking_lot solves. Maybe throw in a TODO here to research this some time in the future.

Reply:
Threw in a TODO.